packages/integrations/src/tracing.ts

```diff
@@ -0,0 +1,170 @@
+import { EventProcessor, Hub, Integration } from '@sentry/types';
+import { fill, getGlobalObject, isMatchingPattern, SentryError, supportsNativeFetch } from '@sentry/utils';
+
+/** JSDoc */
+interface TracingOptions {
+  tracingOrigins: Array<string | RegExp>;
+  traceXHR?: boolean;
+  traceFetch?: boolean;
+  autoStartOnDomReady?: boolean;
+}
+
+/**
+ * Tracing Integration
+ */
+export class Tracing implements Integration {
+  /**
+   * @inheritDoc
+   */
+  public name: string = Tracing.id;
+
+  /**
+   * @inheritDoc
+   */
+  public static id: string = 'Tracing';
+
+  /**
+   * If we have an xhr we need to store the url in the instance.
+   *
+   */
+  // @ts-ignore
+  private _xhrUrl?: string;
+
+  /**
+   * Constructor for Tracing
+   *
+   * @param _options TracingOptions
+   */
+  public constructor(private readonly _options: TracingOptions) {
+    if (!_options.tracingOrigins || !Array.isArray(_options.tracingOrigins) || _options.tracingOrigins.length === 0) {
+      throw new SentryError('You need to define `tracingOrigins` in the options. Set an array of urls to trace.');
+    }
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public setupOnce(_: (callback: EventProcessor) => void, getCurrentHub: () => Hub): void {
+    if (this._options.traceXHR !== false) {
+      this._traceXHR(getCurrentHub);
+    }
+    if (this._options.traceFetch !== false) {
+      this._traceFetch(getCurrentHub);
+    }
+    if (this._options.autoStartOnDomReady !== false) {
+      getGlobalObject<Window>().addEventListener('DOMContentLoaded', () => {
+        Tracing.startTrace(getCurrentHub(), getGlobalObject<Window>().location.href);
+      });
+    }
+  }
+
+  /**
+   * Starts a new trace
+   * @param hub The hub to start the trace on
+   * @param transaction Optional transaction
+   */
+  public static startTrace(hub: Hub, transaction?: string): void {
+    hub.configureScope(scope => {
+      scope.startSpan();
+      scope.setTransaction(transaction);
+    });
+  }
+
+  /**
+   * JSDoc
+   */
+  private _traceXHR(getCurrentHub: () => Hub): void {
+    if (!('XMLHttpRequest' in global)) {
+      return;
+    }
+
+    const xhrproto = XMLHttpRequest.prototype;
+
+    fill(
+      xhrproto,
+      'open',
+      originalOpen =>
+        function(this: XMLHttpRequest, ...args: any[]): void {
+          // @ts-ignore
+          const self = getCurrentHub().getIntegration(Tracing);
+          if (self) {
+            self._xhrUrl = args[1] as string;
+          }
+          // tslint:disable-next-line: no-unsafe-any
+          return originalOpen.apply(this, args);
+        },
+    );
+
+    fill(
+      xhrproto,
+      'send',
+      originalSend =>
+        function(this: XMLHttpRequest, ...args: any[]): void {
+          // @ts-ignore
+          const self = getCurrentHub().getIntegration(Tracing);
```

Comment:
This is very confusing to read. `self` usually means `this` to preserve scope, but in our case, `this` is XHR 😅

Reply:
I know, fill, wrap, this, self, fml.